### PR TITLE
feat(api, spi, storage): policy profile implementation

### PR DESCRIPTION
**Pull request by saschaisele-zf**
## What this PR changes/adds
This PR adds the odrl:profile attribute to the policy model.
This includes the necessary changes to the api, validation, transformation, spi and storage.
**BREAKING CHANGE**: a new `profiles JSON` column has been added to the `edc_policydefinitions` table  

## Why it does that
The change is necessary to conform to ODRL v2.2

## Further notes
For better validation, a new array validator named 'MandatoryIdArray' was added that ensures the array is a list of 'id' values.
To accomodate the transformation, the 'AbstractJsonLdTransformer' was extended with the 'transformIdStringArray' method.

## Linked Issue(s)
Closes [#3691](https://github.com/eclipse-edc/Connector/issues/3961)

Please be sure to take a look at the [contributing guidelines](https://github.com/eclipse-edc/.github/blob/main/CONTRIBUTING.md#submit-a-pull-request) and our [etiquette for pull requests](https://github.com/eclipse-edc/.github/blob/main/contributing/pr_etiquette.md).

### CODE REVIEW

**Comment by saschaisele-zf** [comment]
@ndr-brt

**Comment by paullatzelsperger** [comment]
@saschaisele-zf I am currently working on adding the `/v3.1alpha` version to the management API, because I will need that for another work item too, and I want to avoid duplicated work.

I assume i'll have this done shortly. Could you wait a little, and rebase on that once i'm finished?

**Comment by saschaisele-zf** [comment]
Can do that.
I will just revert the last commit then.

**Comment by paullatzelsperger** [comment]
#4343 is merged, good to rebase

File: extensions/control-plane/api/management-api/policy-definition-api/src/main/java/org/eclipse/edc/connector/controlplane/api/management/policy/v3/PolicyDefinitionApiV3.java [hunk 1]
```diff
@@ -119,6 +119,7 @@ record PolicyDefinitionInputSchema(
                         "@context": "http://www.w3.org/ns/odrl.jsonld",
                         "@type": "Set",
                         "uid": "http://example.com/policy:1010",
+                        "profile": "http://example.com/odrl:profile:02",
```
Comment (ndr-brt) [inline extensions/control-plane/api/management-api/policy-definition-api/src/main/java/org/eclipse/edc/connector/controlplane/api/management/policy/v3/PolicyDefinitionApiV3.java:122]: this should be added in a `PolicyDefinitionApiV31alpha interface, extended by a dedicated controller

File: core/common/lib/validator-lib/src/main/java/org/eclipse/edc/validator/jsonobject/validators/OptionalIdArray.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+    private ValidationResult validateMin(JsonArray array) {
+        if (min == null || (array.size() >= min)) {
+            return ValidationResult.success();
+        }
+        return ValidationResult.failure(violation(format("array '%s' should at least contains '%s' elements", path, min), path.toString()));
+    }
+
+    private ValidationResult validateType(JsonArray array) {
+        for (JsonValue value : array) {
```
Comment (ndr-brt) [inline core/common/lib/validator-lib/src/main/java/org/eclipse/edc/validator/jsonobject/validators/OptionalIdArray.java:74]:
```suggestion
        for (var value : array) {
```

File: core/common/lib/validator-lib/src/main/java/org/eclipse/edc/validator/jsonobject/validators/OptionalIdArray.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            try {
+                var id = value.asJsonObject().getJsonString(ID);
+                if (id == null || id.getString().isBlank() || id.getString().isEmpty()) {
```
Comment (ndr-brt) [inline core/common/lib/validator-lib/src/main/java/org/eclipse/edc/validator/jsonobject/validators/OptionalIdArray.java:77]: the `isEmpty` check is unneeded

File: core/common/lib/validator-lib/src/test/java/org/eclipse/edc/validator/jsonobject/JsonObjectValidatorTest.java [hunk 4]
```diff
@@ -217,6 +218,73 @@ void shouldFail_whenInputIsNull() {
                 });
     }
 
+    @Test
```
Comment (ndr-brt) [inline core/common/lib/validator-lib/src/test/java/org/eclipse/edc/validator/jsonobject/JsonObjectValidatorTest.java:221]: please wrap these tests specific to the `OptionalIdArray` in a `@Nested` dedicated class

File: spi/common/json-ld-spi/src/main/java/org/eclipse/edc/jsonld/spi/transformer/AbstractJsonLdTransformer.java [hunk 5]
```diff
@@ -280,6 +283,60 @@ protected String transformString(JsonValue value, TransformerContext context) {
...
+    /**
+     * Transforms a JsonValue to a list of strings. If the value parameter is not of type JsonString, JsonObject or JsonArray,
+     * a problem is reported to the context and an empty list is returned.
+     *
+     * @param value   the value to transform
+     * @param context the transformer context
+     * @return the string list result
+     */
+    @Nullable
+    protected List<String> transformIdStringArray(JsonValue value, TransformerContext context) {
```
Comment (ndr-brt) [inline spi/common/json-ld-spi/src/main/java/org/eclipse/edc/jsonld/spi/transformer/AbstractJsonLdTransformer.java:295]: this looks overcomplicated, `profile`, when not-null, is always a (non-empty) list, as verified by the validator, so it should be treated as such.
only the code from 315 to 324 can be used (please move it in the calling class, there's no reason to keep it here).
the "isEmpty" validation can be avoided as well

**Review by ndr-brt — CHANGES_REQUESTED** [review]
please add a test on `PolicyDefinitionStoreTestBase` as well, to ensure that the persistence is working correctly

File: extensions/control-plane/api/management-api/policy-definition-api/src/test/java/org/eclipse/edc/connector/controlplane/api/management/policy/v31alpha/PolicyDefinitionApiV31AlphaControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+    @Override
+    protected Object controller() {
+        return new PolicyDefinitionApiV31AlphaController(monitor, transformerRegistry, service, validatorRegistry);
+    }
+
+    @Override
+    protected RequestSpecification baseRequest() {
+        return given()
+                .baseUri("http://localhost:%d/v3/policydefinitions".formatted(port))
```
Comment (ndr-brt) [inline extensions/control-plane/api/management-api/policy-definition-api/src/test/java/org/eclipse/edc/connector/controlplane/api/management/policy/v31alpha/PolicyDefinitionApiV31AlphaControllerTest.java]:
```suggestion
                .baseUri("http://localhost:%d/v3.1alpha/policydefinitions".formatted(port))
```

File: extensions/control-plane/api/management-api/policy-definition-api/src/main/java/org/eclipse/edc/connector/controlplane/api/management/policy/v3/PolicyDefinitionApiV3.java (hunk 1 above)
Comment (saschaisele-zf) [reply extensions/control-plane/api/management-api/policy-definition-api/src/main/java/org/eclipse/edc/connector/controlplane/api/management/policy/v3/PolicyDefinitionApiV3.java]: done

File: core/common/lib/validator-lib/src/main/java/org/eclipse/edc/validator/jsonobject/validators/OptionalIdArray.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+    @Override
+    public ValidationResult validate(JsonObject input) {
+        ValidationResult sizeResult;
+
+        sizeResult = Optional.ofNullable(input.getJsonArray(path.last()))
+                .map(this::validateMin)
+                .orElse(ValidationResult.success());
+
+        if (sizeResult.failed()) return sizeResult;
```
[the remark this answers is not in the document]
Comment (saschaisele-zf) [reply core/common/lib/validator-lib/src/main/java/org/eclipse/edc/validator/jsonobject/validators/OptionalIdArray.java]: done

File: core/common/lib/validator-lib/src/main/java/org/eclipse/edc/validator/jsonobject/validators/OptionalIdArray.java (hunk 2 above)
Comment (saschaisele-zf) [reply core/common/lib/validator-lib/src/main/java/org/eclipse/edc/validator/jsonobject/validators/OptionalIdArray.java]: done

File: core/common/lib/validator-lib/src/main/java/org/eclipse/edc/validator/jsonobject/validators/OptionalIdArray.java (hunk 3 above)
Comment (saschaisele-zf) [reply core/common/lib/validator-lib/src/main/java/org/eclipse/edc/validator/jsonobject/validators/OptionalIdArray.java]: done

File: core/common/lib/validator-lib/src/test/java/org/eclipse/edc/validator/jsonobject/JsonObjectValidatorTest.java (hunk 4 above)
Comment (saschaisele-zf) [reply core/common/lib/validator-lib/src/test/java/org/eclipse/edc/validator/jsonobject/JsonObjectValidatorTest.java]: done

File: spi/common/json-ld-spi/src/main/java/org/eclipse/edc/jsonld/spi/transformer/AbstractJsonLdTransformer.java (hunk 5 above)
Comment (saschaisele-zf) [reply spi/common/json-ld-spi/src/main/java/org/eclipse/edc/jsonld/spi/transformer/AbstractJsonLdTransformer.java]: done

**Review by ndr-brt — APPROVED** [review]

**Pull request merged**